### PR TITLE
Allow RootContext to create child contexts for streams.

### LINKED ISSUE #6: Unlike in other SDKs, there is no link between `RootContext` and `HttpContext` / `StreamContext`
## Summary

* Due to a missing link between `RootContext` and `HttpContext`, it's not possible to to apply to `HttpContext` configuration that was set in the `RootContext` during `proxy_on_configure`
* Why is `HttpContext` gets created by a standalone factory function rather than by `RootContext` ?

### PR DESCRIPTION
**Pull request by dgn**
To fix #6 

Without this, it's impossible (or just incredibly hard, I tried for hours) to read filter config from within an `HttpContext` or `StreamContext`. It requires users to always supply a `RootContext`, which adds a bit of boilerplate, but IMO is a reasonable way to connect `RootContext` with its children in Rust, as there's no easy way to manage global state.

I'm still very new to Rust so this might be the wrong way to do it. Let's discuss.

### CODE REVIEW

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/dispatcher.rs [hunk 1]
```diff
@@ -96,12 +76,9 @@ impl Dispatcher {
...
     fn create_stream_context(&self, context_id: u32, root_context_id: u32) {
-        if !self.roots.borrow().contains_key(&root_context_id) {
-            panic!("invalid root_context_id")
-        }
-        let new_context = match self.new_stream.get() {
-            Some(f) => f(context_id, root_context_id),
-            None => panic!("missing constructor"),
+        let new_context = match self.roots.borrow().get(&root_context_id) {
+            Some(root_context) => root_context.create_stream_context(context_id, root_context_id),
+            None => panic!("invalid root_context_id"),
```
[the remark this answers is not in the document]
Comment (dgn) [reply src/dispatcher.rs]: I moved all context-mgmt/creation into `on_create_context` now

File: src/dispatcher.rs [hunk 2]
```diff
@@ -114,12 +91,9 @@ impl Dispatcher {
...
     fn create_http_context(&self, context_id: u32, root_context_id: u32) {
-        if !self.roots.borrow().contains_key(&root_context_id) {
-            panic!("invalid root_context_id")
-        }
-        let new_context = match self.new_http_stream.get() {
-            Some(f) => f(context_id, root_context_id),
-            None => panic!("missing constructor"),
+        let new_context = match self.roots.borrow().get(&root_context_id) {
+            Some(root_context) => root_context.create_http_context(context_id, root_context_id),
+            None => panic!("invalid root_context_id"),
```
[the remark this answers is not in the document]
Comment (dgn) [reply src/dispatcher.rs]: I moved all context-mgmt/creation into `on_create_context` now

File: src/dispatcher.rs [hunk 3]
```diff
@@ -144,11 +118,15 @@ impl Dispatcher {
...
-            self.create_http_context(context_id, root_context_id);
-        } else if self.new_stream.get().is_some() {
-            self.create_stream_context(context_id, root_context_id);
+            self.create_root_context(context_id);
+        } else if let Some(root_context) = self.roots.borrow().get(&root_context_id) {
+            match root_context.get_type() {
+                ContextType::HttpContext => self.create_http_context(context_id, root_context_id),
+                ContextType::StreamContext => {
+                    self.create_stream_context(context_id, root_context_id)
+                }
```
[the remark this answers is not in the document]
Comment (dgn) [reply src/dispatcher.rs]: agreed. removed it from the trait and the examples

File: src/traits.rs [hunk 4]
```diff
@@ -159,6 +175,16 @@ pub trait StreamContext: Context {
     fn on_log(&mut self) {}
 }
 
+struct EmptyHttpContext;
+
+impl HttpContext for EmptyHttpContext {}
+impl Context for EmptyHttpContext {}
```
[the remark this answers is not in the document]
Comment (dgn) [reply src/traits.rs]: done

File: src/traits.rs [hunk 5]
```diff
@@ -121,6 +121,22 @@ pub trait RootContext: Context {
...
+        &self,
+        _context_id: u32,
+        _root_context_id: u32,
+    ) -> Box<dyn StreamContext> {
+        Box::new(EmptyStreamContext)
+    }
+
+    fn get_type(&self) -> ContextType {
+        ContextType::RootContext
+    }
```
[the remark this answers is not in the document]
Comment (dgn) [reply src/traits.rs]: if I had to choose I'd prefer the former (ie have a single `new_context` fn that creates whatever type), but at this point I think I like the more explicit approach more- because in the end we would have to be type-casting in the SDK again to call trait-specific fn's. but I'm open to suggestions

File: src/dispatcher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply src/dispatcher.rs]: Sorry, I didn't notice this was called from `on_create_context()`, and I thought it was a dead code... I think it was more readable when it was split into separate functions. Could you undo this last update? Thanks and sorry for the noise.

File: examples/http_body.rs [hunk 6]
```diff
@@ -18,10 +18,23 @@ use proxy_wasm::types::*;
...
+    fn get_type(&self) -> ContextType {
+        ContextType::HttpContext
+    }
+
+    fn create_http_context(&self, _context_id: u32) -> Box<dyn HttpContext> {
+        Box::new(HttpBody)
+    }
+}
+
+impl Context for HttpBodyRoot {}
```
Comment (PiotrSikora) [inline examples/http_body.rs:37]: Nit: move `impl Context` before `impl RootContext` to match order for `HttpBody`.

File: src/traits.rs [hunk 7]
```diff
@@ -121,6 +121,18 @@ pub trait RootContext: Context {
...
+        Box::new(EmptyHttpContext)
+    }
+
+    fn create_stream_context(&self, _context_id: u32) -> Box<dyn StreamContext> {
+        Box::new(EmptyStreamContext)
+    }
+
+    fn get_type(&self) -> ContextType {
+        ContextType::RootContext
+    }
```
Comment (PiotrSikora) [inline src/traits.rs:135]: Shouldn't those functions be split into a separate trait?

File: examples/http_auth_random.rs [hunk 8]
```diff
@@ -20,10 +20,23 @@ use std::time::Duration;
 #[no_mangle]
 pub fn _start() {
     proxy_wasm::set_log_level(LogLevel::Trace);
-    proxy_wasm::set_http_context(|_, _| -> Box<dyn HttpContext> { Box::new(HttpAuthRandom) });
+    proxy_wasm::set_root_context(|_| -> Box<dyn RootContext> { Box::new(HttpAuthRandomRoot) });
```
Comment (PiotrSikora) [inline examples/http_auth_random.rs:23]: Looking at this example, I wonder if we couldn't do something type-specific here, e.g.
```
proxy_wasm::set_root_context(|_| -> Box<dyn RootContext> { Box::new(HttpHeadersRoot) });
proxy_wasm::set_http_context(|root_context, context_id| -> Box<dyn HttpContext> {
    root_context.new_http_context(context_id)
});
```
with dispatcher taking care of resolving `root_context_id` to `root_context` before calling this callback?

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

**Review by PiotrSikora — COMMENTED** [review]
Also, could you merge/rebase on top of current master branch? Thanks!

File: examples/http_auth_random.rs (hunk 8 above)
Comment (PiotrSikora) [reply examples/http_auth_random.rs:23]: On the second thought, that might be useless, since we cannot expose multiple interfaces right now anyway.

File: src/dispatcher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dgn) [reply src/dispatcher.rs]: reverted

File: src/traits.rs (hunk 7 above)
Comment (dgn) [reply src/traits.rs:135]: to avoid the need for a `RootContext`? I'm not sure about that. I think `RootContext` (by its name) already has a notion of controlling its child contexts, so this is what I would expect, instead of another meta-struct that creates all contexts.

Also, the main reason to implement this was so that the `on_configure` calls could be passed on to child contexts easily by the `RootContext`. if we move context creation out of the root context again, it'll again lose the ability to map `RootContext` to e.g. `HttpContext` - except if the user explicitly keeps a mapping in that struct

File: examples/http_auth_random.rs [hunk 9]
```diff
@@ -20,31 +20,21 @@ use std::time::Duration;
...
 }
 
 struct HttpAuthRandom;
```
Comment (PiotrSikora) [inline examples/http_auth_random.rs:26]: Nit: move `struct HttpAuthRandom;` after definition of `HttpAuthRandomRoot`.

File: examples/http_body.rs [hunk 10]
```diff
@@ -18,10 +18,23 @@ use proxy_wasm::types::*;
 #[no_mangle]
 pub fn _start() {
     proxy_wasm::set_log_level(LogLevel::Trace);
-    proxy_wasm::set_http_context(|_, _| -> Box<dyn HttpContext> { Box::new(HttpBody) });
+    proxy_wasm::set_root_context(|_| -> Box<dyn RootContext> { Box::new(HttpBodyRoot) });
 }
 
 struct HttpBody;
```
Comment (PiotrSikora) [inline examples/http_body.rs:24]: Nit: move `struct HttpBody;` after definition of `HttpBodyRoot`.

**Review by PiotrSikora — COMMENTED** [review]
Thanks! Could you also add an example showing developers how to use this feature? Maybe add a new header which value is passed via configuration to `examples/http_headers.rs`?

Note to self: merge this after tagging v0.1.3 (pending fix for #40).

File: src/types.rs [hunk 11]
```diff
@@ -85,3 +85,9 @@ pub enum MetricType {
 }
 
 pub type Bytes = Vec<u8>;
+
+pub enum ContextType {
```
Comment (PiotrSikora) [inline src/types.rs:89]: Please add:
```
#[repr(u32)]
#[derive(Debug)]
```
(it's not used in the ABI right now, but it's going to be soon, so we might as well define it properly now)

Also, could you move it below `Status`?

File: src/dispatcher.rs [hunk 12]
```diff
@@ -131,26 +116,19 @@ impl Dispatcher {
...
+                ContextType::RootContext => panic!("missing ContextType on root_context"),
+            }
```
Comment (PiotrSikora) [inline src/dispatcher.rs:129]: One more thing. I think it would be useful if we could retain backwards compatibility, and merge this without breaking all existing plugins, even if there aren't too many in the wild right now.

One way to do this is to fallback to the standalone constructors in case `root_context.get_type()` returns `ContextType::RootContext` or `None`, and perhaps revert changes to `examples/http_auth_random.rs` to make sure that the old code path still works.

File: src/traits.rs [hunk 13]
```diff
@@ -121,6 +121,18 @@ pub trait RootContext: Context {
     fn on_queue_ready(&mut self, _queue_id: u32) {}
 
     fn on_log(&mut self) {}
+
+    fn create_http_context(&self, _context_id: u32) -> Box<dyn HttpContext> {
+        Box::new(EmptyHttpContext)
```
Comment (PiotrSikora) [inline src/traits.rs:126]: `Box<dyn HttpContext>` -> `Option<Box<dyn HttpContext>>` and return `None`.

File: src/traits.rs [hunk 14]
```diff
@@ -121,6 +121,18 @@ pub trait RootContext: Context {
...
+    }
+
+    fn create_stream_context(&self, _context_id: u32) -> Box<dyn StreamContext> {
+        Box::new(EmptyStreamContext)
```
Comment (PiotrSikora) [inline src/traits.rs:130]: Ditto.

File: src/traits.rs [hunk 15]
```diff
@@ -121,6 +121,18 @@ pub trait RootContext: Context {
...
+    }
+
+    fn get_type(&self) -> ContextType {
+        ContextType::RootContext
```
Comment (PiotrSikora) [inline src/traits.rs:134]: `Option<ContextType>` and return `None`.

File: src/traits.rs [hunk 16]
```diff
@@ -159,6 +171,11 @@ pub trait StreamContext: Context {
     fn on_log(&mut self) {}
 }
 
+struct EmptyStreamContext;
+
+impl StreamContext for EmptyStreamContext {}
+impl Context for EmptyStreamContext {}
```
Comment (PiotrSikora) [inline src/traits.rs:177]: This won't be needed with `Option<T>`.

File: src/traits.rs [hunk 17]
```diff
@@ -303,3 +320,8 @@ pub trait HttpContext: Context {
 
     fn on_log(&mut self) {}
 }
+
+struct EmptyHttpContext;
+
+impl HttpContext for EmptyHttpContext {}
+impl Context for EmptyHttpContext {}
```
Comment (PiotrSikora) [inline src/traits.rs:327]: This won't be needed with `Option<T>`.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/dispatcher.rs (hunk 12 above)
Comment (dgn) [reply src/dispatcher.rs:129]: +1 to remain backwards-compatible, but I think we should then deprecate the old way to create contexts and remove it in 0.2.0

Comment (dgn) [reply src/dispatcher.rs:129]: that would also allow us to remove `NoopRoot` in 0.2.0

File: examples/http_body.rs (hunk 6 above)
Comment (dgn) [reply examples/http_body.rs]: done

File: src/types.rs (hunk 11 above)
Comment (dgn) [reply src/types.rs]: done

File: src/traits.rs (hunk 13 above)
Comment (dgn) [reply src/traits.rs]: done

File: src/traits.rs (hunk 14 above)
Comment (dgn) [reply src/traits.rs]: done

File: src/traits.rs (hunk 16 above)
Comment (dgn) [reply src/traits.rs]: done

File: src/traits.rs (hunk 17 above)
Comment (dgn) [reply src/traits.rs]: removed

File: examples/http_config.rs [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+impl HttpContext for HttpConfigHeader {
+    fn on_http_response_headers(&mut self, _num_headers: usize) -> Action {
+        self.add_http_response_header("custom-header", self.header_content.as_str());
+
+        Action::Continue
+    }
+}
+
+struct HttpConfigHeaderRootContext {
```
Comment (PiotrSikora) [inline examples/http_config.rs:43]: Nit: name this `HttpConfigHeaderRoot` and move it before `HttpConfigHeader` to match style used in other examples.

File: examples/http_auth_random.rs [hunk 19]
```diff
@@ -25,6 +25,24 @@ pub fn _start() {
...
+        }
+        trace!("Access forbidden.");
+        self.send_http_response(
+            403,
+            vec![("Powered-By", "proxy-wasm")],
+            Some(b"Access forbidden.\n"),
+        );
+    }
+}
+
```
Comment (PiotrSikora) [inline examples/http_auth_random.rs:45]: Nit: this is unrelated and unnecessary move, please undo.

File: examples/http_config.rs [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    fn on_configure(&mut self, _plugin_configuration_size: usize) -> bool {
+        if let Some(config_bytes) = self.get_configuration() {
+            self.header_content = str::from_utf8(config_bytes.as_ref()).unwrap().to_owned()
+        }
+        true
+    }
+
+    fn create_http_context(&self, _context_id: u32) -> Option<Box<dyn HttpContext>> {
+        Some(Box::new(HttpConfigHeader {
+            header_content: self.header_content.clone(),
```
Comment (PiotrSikora) [inline examples/http_config.rs:59]: This seems expensive, since presumably each context will own a copy of the header value.

It's fine in this example, since the value is small, but in general, wouldn't it be better to store a reference to the parent context and access configuration from there?

File: src/dispatcher.rs (hunk 12 above)
Comment (PiotrSikora) [reply src/dispatcher.rs:167]: The "old way" is fine for plugins that don't have configuration (e.g. where everything is already in the code), and the "new way" can double the code size with unnecessary boilerplate.

But it's something to consider for sure.

File: src/types.rs [hunk 21]
```diff
@@ -47,6 +47,13 @@ pub enum Status {
     InternalFailure = 10,
 }
 
+#[repr(u32)]
+#[derive(Debug)]
+pub enum ContextType {
+    HttpContext = 0,
```
Comment (PiotrSikora) [inline src/types.rs:53]: Could you re-add `RootContext` as `0` here? Thanks.

Comment (dgn) [reply src/types.rs:53]: Why? How would that be different from returning `None`?

**Review by PiotrSikora — APPROVED** [review]
